blockchain/blockdao/blockdao.go

```diff
@@ -174,9 +177,53 @@ func (dao *blockDAO) Start(ctx context.Context) error {
 			return errors.Wrap(err, "failed to write initial value for top height")
 		}
 	}
-	return dao.initStores()
+	err = dao.initStores()
+	if err != nil {
+		return err
+	}
+	return dao.adjust()
+}
+func (dao *blockDAO) adjust() error {
+	topHeight, err := dao.getTipHeight()
+	if err != nil {
+		return err
+	}
+	topDB, index, err := dao.getTopDB(topHeight)
+	if err != nil {
+		return err
+	}
+	if index == 0 {
+		return nil
+	}
+	value, err := topDB.Get(blockNS, topHeightKey)
+	if err != nil {
+		return err
+	}
+	topHeightOfBlockDB := enc.MachineEndian.Uint64(value)
+	if topHeight == topHeightOfBlockDB {
+		return nil
+	} else if topHeight > topHeightOfBlockDB {
+		return ErrMissingBlock
+	}
+	return dao.rebuild(topHeight+1, topHeightOfBlockDB, topDB)
+}
+func (dao *blockDAO) rebuild(from, to uint64, store db.KVStore) (err error) {
+	batch := db.NewBatch()
+	for i := from; i <= to; i++ {
+		heightValue := byteutil.Uint64ToBytes(i)
+		heightKey := append(heightPrefix, heightValue...)
+		hash, err := store.Get(blockHashHeightMappingNS, heightKey)
```

Comment:
this is correct in the assumption that block "from" is in "store". If "from" is not in "store", the recovery will fail. It is acceptable, but not ideal.